### ISSUE TITLE
Duplicate packages between bioconda and conda-forge

### ISSUE THREAD

**Issue by keuv-grvl** [bioconda/bioconda-recipes#2796]
Bioconda relies on conda-forge which is a general purpose channel. There are some (105) duplicates between bioconda and conda-forge. Here is the list: https://gist.github.com/keuv-grvl/1580f12a07d400728ec690ec517229a8
Maintaining several recipes for the same software will be counterproductive as we can freely submit PR to `conda-forge` or `bioconda`. Moreover, the conda-forge community is active.

Possible solutions could be:
1. Simply drop duplicates from bioconda
1. Move duplicates to a distinct git repo (`bioconda-deprecated-recipes`?)
1. Move duplicates to another folder within [bioconda-recipes](https://github.com/bioconda/bioconda-recipes)
1. Set higher priority for `conda-forge` than `bioconda` during build

# Some pros and cons for each

## 1
- Some tools won't behave exactly the same (eg: #2554)
- All packages have to be rebuild?
- High risk of breaking various packages?
## 2
- We keep track of deprecated recipes
- We don't "pollute" this repo
- Can we build them if needed?
## 3
- We keep track of deprecated recipes
- We could build them
- Wa can set `build: string: deprecated` for user information
## 4
- `conda-forge` packages will be used over bioconda packages (#2554)
- We can't override package if needed but issues and PR to conda-forge are possible

I can't see which solution is better. Any thoughts @bioconda/core ?

**Comment by daler** [comment]
@keuv-grvl thanks for looking into this. It would be helpful to have this duplicates-checking built in to bioconda-utils. Perhaps a way that pays attention to versions and platforms available might be to compare the packages in each channel's `https://conda.anaconda.org/{channel}/{arch}/repodata.json`.

Channel priority would work in theory, but we've been playing around with that. Since we also have to deal with the `r` channel, we actually decided conda-forge should be lowest priority (see #2675).

Moving recipes to a separate dir in the repo or another repo altogether wouldn't help much because the packages will still exist on anaconda, available for building and installing.

I think the only long-term solution is to delete the duplicates from the bioconda channel on anaconda and rebuild the bioconda packages that were dependent on them. This would take some planning and careful testing. We'd have to set up a temporary channel that has the duplicates removed, and at build time replace the bioconda channel with this temp channel to make sure that the conda-forge packages work as drop-in replacements. If there are a large number of packages to be rebuilt, we could inspect the current DAG of bioconda recipes to see if there are sub-dags that contain a smaller set of duplicates that can be worked on rather than everything at once.

**Comment by bgruening** [comment]
An other option is to put all these packages on a blacklist with a comment that these are deprecated and conda-forge is the way to go. Disadvantage is that they are still visible to our contributors and maybe someone updates them. We could also write a big disclaimer in meta.yml file ...?

Having a separate folder seems als a very nice idea, something like `deprecated` but this only makes sense if we restructure our repository and have a small number of sub folders.

> I think the only long-term solution is to delete the duplicates from the bioconda channel on anaconda

I don't think this is really feasible, rebuilding old packages seem like very very hard in bioconda ... at least with the exact package depended package version. We sill hide old packages in the git history.

**Comment by keuv-grvl** [comment]
Just to be sure I understand. If a package is blacklisted, it won't be used as a build dependency? So the `conda-forge` (or `r`, or `defaults`) package will be used even if versions are the same and bioconda have higher priority ? That sounds great.


A disclaimer could be added in each `meta.yaml` like:
```
build:
  number: +1
  string: deprecated
```
The package name will be `mypackage-1.2.3-deprecated`.

**Comment by daler** [comment]
Blacklists are only for determining which recipes (i.e., directories containing a meta.yaml) to build on travis-ci. It doesn't have any say about installing dependencies for building or installing. That is determined by the channels configured and which packages (i.e., name-version-build.tar.bz2) are available on anaconda.org for each channel.

So unfortunately the blacklist won't work as a mechanism for fine-tuning what dependencies should be considered.

**Comment by bgruening** [comment]
@daler I'm just talking about the best strategy to reitere a package on github not during the installation. Currently, my only concern is that we should tell our developers to __not__ use this recipe anymore, but please go to conda-forge and upgrade it.

**Comment by daler** [comment]
@bgruening ah, I see. I agree with what you said: blacklisting might be confusing, and a deprecated dir prob makes the most sense. We might be able to do it before restructuring by adding a top-level `recipes-deprecated` dir outside of `recipes`.

**Comment by martin-mann** [comment]
a problem I see: *time* .. it takes quite a while till a pull request to conda is finally merged and available as a dependency. So by pushing the stuff to conda (what I would prefer too), you very much delay (and annoy) contributors, since they have to wait until the new version is available before they can continue to setup their package..

how difficult is it to do both (adding to bioconda plus opening a pull request in conda) and to remove the bioconda package as soon as conda has consumed the pull request?

I agree this sounds ugly, but gets the people working on issues rather than waiting on merges..

**Comment by bgruening** [comment]
@daler do we have a plan here? Should we write a script to determine which packages are in conda-forge and move these into a deprecated folder, but leave them in the bioconda channel for now?

**Comment by daler** [comment]
No existing plan, but yours sounds good. I hesitate to delete packages in the name of reproducibility, but shuffling recipes around seems fine.

But . . . how should we determine equality of packages? Sometimes a bioconda package might be intended to replace a conda-forge one, in which case the versions might be the same but the build.sh might be different. I guess we can see how many packages are shared between the channels and that have different meta.yaml or build.sh or patches just to see what we're working with. I'll add a bioconda-utils issue for this.

**Comment by ostrokach** [comment]
> I think the only long-term solution is to delete the duplicates from the bioconda channel on anaconda

I agree that in the long term, we would need some way of removing packages from bioconda while making sure that we don't break any dependencies. I stumbled on this trying to remove `retrying` #4689 because it does not link correctly but still has precendence over the conda-forge version because of the higher build number.

**Comment by bgruening** [comment]
@ostrokach if there is the same version in conda-forge we can move these packages to bioconda-legacy. But I would like to have a conda-forge version before. You can pin a specific verison of `retrying`, do you? Also in your special case we might just move the broken version away.

**Comment by ostrokach** [comment]
Yes, there is the same version of retrying (1.3.3) on conda forge: https://github.com/conda-forge/retrying-feedstock/blob/master/recipe/meta.yaml. I'm the one who added retrying to bioconda, so I'm the one to blame for the bad package. But I figured that since it's in conda-forge already, and since it does not have much to do with bioinformatics, it's best just to remote it.

Thanks for your help @bgruening!

**Comment by kain88-de** [comment]
Removing the duplicates would be nice. Especially since the conda forge recipes tend to be updated more frequently. I recently added bioconda for a single package and this lead to several others now being downgraded.

**Comment by AforAnonymous** [comment]
Does anyone still work on this issue?

**Comment by vinjana** [comment]
I think removing duplicates is possible but should be guided by what bioconda promises in the [docs](https://bioconda.github.io/faqs.html#what-s-the-lifecycle-of-a-bioconda-package), namely that old packages stay in the channel! Without this condition old YAML environment dumps will certainly fail and even worse it may not be possible to reconstitute them exactly (including R, Perl, Python versions?) from packages in other channels.  In the worst case this breaks exact reproducibility of analyses. 

We promoted bioconda a lot in our department, but now we got the first disappointed user, that -- in the middle of a migration of multiple workflows to conda -- found that packages are suddenly missing or that dependencies have changed.

**Comment by daler** [comment]
@vinjana the ecosystem is complicated and moving fast; conda itself is evolving, and there are many parts out of our direct control. We need to somehow strike the balance between halting all work on anything that uses conda and maintaining long-term reproducibility.

Yes, this is a frustrating problem. It's also unfortunately a hard problem with no good solution yet. 

One method we currently use is to move packages from the `main` label, which is available by default, to different labels that need to be explicitly specified. For example https://anaconda.org/bioconda/repo?label=broken or https://anaconda.org/bioconda/repo?label=deprecated. These are removed from the main channel to avoid breaking the majority of environments, but are still available by adding an argument like `--channel bioconda/labels/deprecated` when installing.

If you have environment yaml files that are problematic, please submit an issue that includes them and we'll figure out a solution.

**Comment by vinjana** [comment]
@daler Thanks for the clarifications. For the moment we will use conda only for setting up environments initially and live without any guarantees (in particular for the older workflows). We were considering setting up our own -- sort of -- "long-term storage proxy channel" (or "channel proxy" :-) ) containing only package copies. Something that would allow us to guarantee the reproducibility of at least our own conda environments.

**Comment by bgruening** [comment]
@vinjana Conda has its limitations and we try to fix them / workaround them in various ways. We have written a small manuscript about it here https://www.biorxiv.org/content/early/2017/10/11/200683.

Essentially, we freeze environments into containers (biocontainers, docker and singularity atm) to store them easily and reuse them at any time.
We are also working on a backup for all packages (meaning the package tarball), but also all source tarballs that where used to build the packages.

Let me know if you want to help with any of these projects.

**Comment by no-identd** [comment]
Thanks to @nixos-discourse, I've noticed this thread by @mjlbach over in the NixOS Discourse:

https://discourse.nixos.org/t/installing-python-modules-from-pip-buildfhsuserenv-for-python-tutorial/5704/3

Essentially the same issue as the issue here, except on a lot of Steroids, since it involves nightlies and must deal with the "no non-pure python modules" restriction on `pip install` in NixOS. Anyone got any ideas? (Technically I suppose this would belong in a separate issue, however, I figured I might as well use this opportunity to bump this issue. :p)
